pkg/sync/sync.go

```diff
@@ -176,96 +179,151 @@ var bufPool = sync.Pool{
 	},
 }
 
-func copyObject(src, dst object.ObjectStorage, obj object.Object) error {
-	if limiter != nil {
-		limiter.Wait(obj.Size())
-	}
-	concurrent <- 1
-	defer func() {
-		<-concurrent
-	}()
-	key := obj.Key()
-	in, e := src.Get(key, 0, -1)
-	if e != nil {
-		if _, err := src.Head(key); err != nil {
-			return nil
+func try(n int, f func() error) (err error) {
+	for i := 0; i < n; i++ {
+		err = f()
+		if err == nil {
+			return
 		}
-		return e
+		time.Sleep(time.Second * time.Duration(i*i))
 	}
-	defer in.Close()
+	return
+}
 
-	if obj.Size() <= maxBlock ||
-		strings.HasPrefix(src.String(), "file://") ||
-		strings.HasPrefix(dst.String(), "file://") {
-		return dst.Put(key, in)
+func deleteObj(storage object.ObjectStorage, key string, dry bool) {
+	if dry {
+		logger.Infof("Will delete %s from %s", key, storage)
+		return
+	}
+	start := time.Now()
+	if err := try(3, func() error { return storage.Delete(key) }); err == nil {
+		deleted.Increment()
+		logger.Debugf("Deleted %s from %s in %s", key, storage, time.Since(start))
+	} else {
+		failed.Increment()
+		logger.Errorf("Failed to delete %s from %s in %s: %s", key, storage, time.Since(start), err)
+	}
+}
+
+func copyPerms(dst object.ObjectStorage, obj object.Object) {
+	start := time.Now()
+	key := obj.Key()
+	fi := obj.(object.File)
+	if err := dst.(object.FileSystem).Chmod(key, fi.Mode()); err != nil {
+		logger.Warnf("Chmod %s to %d: %s", key, fi.Mode(), err)
 	}
-	// obj.Size > maxBlock, download the object into disk first
-	f, e := ioutil.TempFile("", "rep")
-	if e != nil {
-		return e
+	if err := dst.(object.FileSystem).Chown(key, fi.Owner(), fi.Group()); err != nil {
+		logger.Warnf("Chown %s to (%s,%s): %s", key, fi.Owner(), fi.Group(), err)
 	}
-	os.Remove(f.Name()) // will be deleted after Close()
-	defer f.Close()
+	logger.Debugf("Copied permissions (%s:%s:%s) for %s in %s", fi.Owner(), fi.Group(), fi.Mode(), key, time.Since(start))
+}
 
+func doCheckSum(src, dst object.ObjectStorage, key string, equal *bool) error {
 	buf := bufPool.Get().(*[]byte)
 	defer bufPool.Put(buf)
-	if _, e = io.CopyBuffer(f, in, *buf); e != nil {
-		return e
+
+	srcIn, err := src.Get(key, 0, -1)
+	if err != nil {
+		return fmt.Errorf("src get: %s", err)
+	}
+	defer srcIn.Close()
+	srcH := md5.New()
+	if _, err = io.CopyBuffer(srcH, srcIn, *buf); err != nil {
```

Comment:
we read from two streams in the same time, should we just compare the bytes?

Or, we can compare the checksum block by block to find the different earlier.

Reply:
For larger objects, we can also compare them in parallel.

Reply:
Parallelism is not implemented because Get() returns a reader, not readerat

Reply:
Get(key, offset, limit)